File: 36.valid-sudoku.cpp
/*
 * @lc app=leetcode id=36 lang=cpp
 *
 * [36] Valid Sudoku
 */

// @lc code=start
class Solution
{
public:
    bool isValidSudoku(vector<vector<char>> &board)
    {

        map<int, set<int>> row_map;
        map<int, set<int>> col_map;
        map<pair<int, int>, set<int>> square_map;

        // lets fill the values
        for (int m = 0; m < board.size(); m++) // row
        {
            for (int n = 0; n < board[0].size(); n++) // col
            {
                if (board[m][n] != '.')
                {
                    // check if it is allready  in the set
                    if (row_map[m].find(board[m][n]) == row_map[m].end())
                    {
                        row_map[m].insert(board[m][n]);
                    }
                    else
                    {
                        return false;
                    }

                    if (col_map[n].find(board[m][n]) == col_map[n].end())
                    {
                        col_map[n].insert(board[m][n]);
                    }
                    else
                    {
                        return false;
                    }

                    if (square_map[{m / 3, n / 3}].find(board[m][n]) == square_map[{m / 3, n / 3}].end())
                    {
                        square_map[{m / 3, n / 3}].insert(board[m][n]);
                    }
                    else
                    {
                        return false;
                    }
                }
            }
        }
        return true;
    }
};
// @lc code=end

File: 36.valid-sudoku.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#

# @lc code=start
class Solution:


    def isValidSudoku(self, board: List[List[str]]) -> bool:

        # this is an O(3n) solution, which reduches to O(N) it uses dictionarys with a lookup of O(1) most of the time in python. and it uses sets whiich have a
        # lookup of O 1 

        col_dict = {} # 0-8
        row_dict = {} # 0-8
        squares_dict = {} # 0-8

        for i in range(9):
            for j in range(9):
                if j not in col_dict.keys():
                    col_dict[j] = set()
                if i not in row_dict.keys():
                    row_dict[i] = set()
                if (i//3,j//3) not in squares_dict.keys():
                    squares_dict[(i//3,j//3)] = set()
                
                if board[i][j] == ".":
                    continue

                
                if board[i][j] in col_dict[j]:
                    return False
                else:
                    col_dict[j].append(board[i][j])

                if board[i][j] in row_dict[i]:
                    return False
                else:
                    row_dict[i].append(board[i][j])

                if board[i][j] in squares_dict[(i//3,j//3)]:
                    return False
                else:
                    squares_dict[(i//3,j//3)].append(board[i][j])
        
        return True
                
                




    
        
# @lc code=end

